### PR TITLE
Add happy-eyeballs device

### PR DESCRIPTION
**Pull request by dinosaure**
Finally, this is the `happy-eyeballs` device (/cc @hannesm). I have mainly one question about the timestamp, currently I gives `0L` but I'm not sure if:
- we should let the user to pass a value directly into the `config.ml` or as a command-line argument
- or if we can call the `Mclock` implementation to give to us the timestamp

### CODE REVIEW

**Comment by hannesm** [comment]
Oops, looks like the API should change. Indeed, the `connect_device` should use the C.elapsed_ns, and not expect a timestamp. I'll prepare a fresh release.

**Comment by hannesm** [comment]
see https://github.com/ocaml/opam-repository/pull/20973

File: lib/mirage/impl/mirage_impl_happy_eyeballs.ml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+open Mirage_impl_dns
+
+type happy_eyeballs = Happy_eyeballs
+
+let happy_eyeballs = Type.v Happy_eyeballs
+
+let generic_happy_eyeballs aaaa_timeout v6_connect_timeout connect_timeout
+    resolve_timeout resolve_retries timer_interval =
+  let packages =
+    [ package "happy-eyeballs-mirage" ~min:"0.2.0" ~max:"1.0.0" ]
```
Comment (hannesm) [inline lib/mirage/impl/mirage_impl_happy_eyeballs.ml:14]:
```suggestion
    [ package "happy-eyeballs-mirage" ~min:"0.3.0" ~max:"1.0.0" ]
```

File: lib/mirage/impl/mirage_impl_happy_eyeballs.ml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+          aaaa_timeout
+          (pp_optional_argument ~name:"v6_connect_timeout")
+          v6_connect_timeout
+          (pp_optional_argument ~name:"connect_timeout")
+          connect_timeout
+          (pp_optional_argument ~name:"resolve_timeout")
+          resolve_timeout
+          (pp_optional_argument ~name:"resolve_retries")
+          resolve_retries
+          (pp_optional_argument ~name:"timer_internal")
```
Comment (hannesm) [inline lib/mirage/impl/mirage_impl_happy_eyeballs.ml:42]:
```suggestion
          (pp_optional_argument ~name:"timer_interval")
```

File: lib/mirage/impl/mirage_impl_happy_eyeballs.ml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    |> cons_if_some timer_interval
+    |> List.map Key.v
+  in
+  let connect _info modname = function
+    | [ _time; _mclock; stack; dns ] ->
+        let pp_optional_argument ~name ppf = function
+          | None -> ()
+          | Some key -> Fmt.pf ppf "?%s:%a " name Key.serialize_call (Key.v key)
+        in
+        Fmt.str {ocaml|%s.connect_device %a%a%a%a%a%a 0L %s %s"|ocaml} modname
```
Comment (hannesm) [inline lib/mirage/impl/mirage_impl_happy_eyeballs.ml:31]:
```suggestion
        Fmt.str {ocaml|%s.connect_device %a%a%a%a%a%a %s %s"|ocaml} modname
```

**Comment by dinosaure** [comment]
This PR seems fine for you?

**Comment by hannesm** [comment]
yes

**Comment by hannesm** [comment]
this looks fine with me, thanks for your work on that :)

**Pull request merged**